File: EXE1_Java/Q11.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Q11 {
    static int intPow(int x, int pow) {
        int ret = 1;
        while (pow != 0) {
            if ((pow & 1) == 1)
                ret *= x;
            x *= x;
            pow >>= 1;
        }
        return ret;
    }

    static int map(int a, int b) {
        return intPow(2, a) * intPow(3, b);
    }

    static boolean f(int[] count, int n, int a, int minB) {
        // choose [a] and [minB..(until too large)] for this number
        // and pass [0..a-1] and [b+1] for next number
        // return true/false of "should we continue?"

        int b = minB;
        int k = map(a, b);
        while (k < n) {
            for (int nextA = a - 1; nextA >= 0; nextA--) {
                f(count, n - k, nextA, b + 1);
            }
            b++;
            k = map(a, b);
        }
        if (k == n) count[0]++;
        return b != 0;
    }

    static int Q11_2(int n) {
        // instead of finding how many ways,
        // we build the number from 0,
        // add one step by one step,
        // until the number is larger than the target

        // condition 1 is important: None of the summands can divide any of the other summands.

        // consider 2^a * 3^b and 2^c * 3^d
        // if a >= c && b >= d, then condition 1 must violate
        // question: is diff set of (a, b) map to unique numer? ans: yes
        int[] count = new int[1];
        int a = 0;
        while (f(count, n, a, 0)) a++;

        return count[0];
    }

    static void Q11(List<String> sb) {
        System.out.println(Q11_2(Integer.parseInt(sb.get(0))));
    }

    public static void main(String[] args) {
        BufferedReader sr = new BufferedReader(new InputStreamReader(System.in));
        String s = null;
        do {
            try {
                List<String> sb = new ArrayList<>();
                while ((s = sr.readLine()) != null) {
                    if (s.length() != 0) sb.add(s);
                    else if (sb.size() != 0) break;
                }
                Q11(sb);
            } catch (Exception e) {
                e.printStackTrace();
            }
        } while (s != null);
    }
}

File: EXE1_Python/Q11.py
#!/usr/bin/env python3

import sys
import traceback

def mapped(a, b):
    return 2**a * 3**b

class RES:
    def __init__(self):
        self.count = 0

def f(res, n, a, minB):
    # choose [a] and [minB..(until too large)] for this number
    # and pass [0..a-1] and [b+1] for next number
    # return true/false of "should we continue?"

    b = minB
    k = mapped(a, b)
    while k < n:
        nextA = a - 1
        while nextA >= 0:
            f(res, n - k, nextA, b + 1)
            nextA -= 1
        b += 1
        k = mapped(a, b)
    if k == n:
        res.count += 1
    return b != 0

def Q11(sb):
    # instead of finding how many ways,
    # we build the number from 0,
    # add one step by one step,
    # until the number is larger than the target

    # condition 1 is important: None of the summands can divide any of the other summands.

    # consider 2^a * 3^b and 2^c * 3^d
    # if a >= c && b >= d, then condition 1 must violate
    # question: is diff set of (a, b) map to unique numer? ans: yes
    n = int(sb[0])

    res = RES()
    a = 0
    while f(res, n, a, 0):
        a += 1
    print(res.count)


if __name__ == '__main__':
    """Verify test cases easily
    1. run the program
    2. paste test case to stdin
    3. press double Enter
    4. continue pasting next test case
    """
    do_next = True
    while do_next:
        try:
            do_next = False  # do_next won't set if encounter EOF
            sb = []
            for line in sys.stdin:  # s = data + '\n'
                line = line.rstrip('\r\n')
                if line:  # check str not empty
                    sb.append(line)
                elif sb:  # check list not empty
                    do_next = True
                    break
            Q11(sb)
        except Exception:
            traceback.print_exc()
